Salesforce Connector Additional Config Topic Edit (#622)
mulesoft/docs-connectors: modules/ROOT/pages/salesforce/salesforce-connector-100-config-topics.adoc

> ifndef::env-site,env-github[]

Reviewer: Instructions for edit:

The Salesforce Connector Additional Configuration Topics topic needs a standard edit. I rewrote a lot of the topic, but I think there are still some awkward areas. I may not be able to put in all of your suggestions, but I'd like to capture them in this editing pass. Here is the PR: https://github.com/mulesoft/docs-connectors/pull/622/
If possible, please complete the edit by or on Wednesday.

include::_attributes.adoc[]
endif::[]

> Read this topic to obtain additional configuration and usage information for the Anypoint Connector for Salesforce (Salesforce Connector).

Reviewer: information for > information about

Reviewer: Watch out for "the" creep:
```suggestion
Read this topic to obtain additional configuration and usage information about Anypoint Connector for Salesforce (Salesforce Connector).
```

Author: I don't agree with using "about."

[[customevents]]
== Custom Event Notifications

Salesforce Connector enables you to obtain custom event notifications. These notifications apply to general events that are not tied to Salesforce data changes.

To obtain custom event notifications:

. Create a streaming channel by using the Publish streaming channel operation.
+
StreamingChannel is a special Salesforce object that represents a channel used for notifying listeners of generic Streaming API events.
+
You can also create a streaming channel through Salesforce or https://workbench.developerforce.com/about.php[

> workbench].

Reviewer: ```suggestion
Workbench].
```

+
. Subscribe to the channel by using the Subscribe channel listener operation.
+
Salesforce Connector converts the custom events in your streaming channel to Mule events.

> For more information about working with streaming channels, see <<streamingchannel,Create a Streaming Channel to Receive Data From Salesforce>> and <<receivedata,Receive Data From Salesforce>>.

Reviewer: ```suggestion
For more information about working with streaming channels, see <<streamingchannel,Create a Streaming Channel to Receive Data from Salesforce>> and <<receivedata,Receive Data from Salesforce>>.
```

[[eventsandtopics]]
== Events and Topics

Your app can receive events by subscribing to an existing Salesforce topic or by creating a new topic and then subscribing to it.

Each event that travels through your flow contains information about the Salesforce data that changed, including how the data changed and when the change occurred.

> Salesforce stores events for 24 hours. A subscriber to an topic or channel can retrieve events related to that topic or channel during the 24 hour retention window. After the retention window ends, the subscriber can retrieve newer events that have not yet expired.

Reviewer: ```suggestion
Salesforce stores events for 24 hours. A subscriber to a topic or channel can retrieve events related to that topic or channel during the 24-hour retention window. After the retention window ends, the subscriber can retrieve newer events that have not yet expired.
```

> Salesforce assigns each broadcast event a numeric ID. IDs are incremented and not guaranteed to be contiguous for consecutive events. For example, the event following the event with ID 999 can have an ID of 1025. A broadcast event ID is unique for the organization and channel. Salesforce does not reuse the IDs of deleted events.

Reviewer: I wonder if it would be simpler to say something like "IDs are incremented, but not necessarily by 1 for each subsequent event."

> For information on processing events when streaming data to an app from Salesforce, see <<Receive Data From Salesforce>>.

Reviewer: information on > information about

Reviewer: ```suggestion
For information on processing events when streaming data to an app from Salesforce, see <<Receive Data from Salesforce>>.
```

[[eventsreceive]]
=== Receive Events for a Topic

> Before you can receive events for Salesforce changes associated with a topic, you must create the topic if it does not exist. When you create a topic, the connector creates a PushTopic, which is a special object in Salesforce that binds a name (in this case, the topic's name) and Salesforce Object Query Language (SOQL) query together. After you create a topic, you can subscribe to it by name.

Reviewer: ```suggestion
Before you can receive events for Salesforce changes associated with a topic, you must create the topic, if it does not exist. When you create a topic, the connector creates a PushTopic, which is a special object in Salesforce that binds a name (in this case, the topic's name) and a Salesforce Object Query Language (SOQL) query together. After you create a topic, you can subscribe to it by name.
```

</flow>
----

> You can subscribe to a topic that was not previously published in Salesforce. However, when the topic is published, you will not receive notifications regarding that topic unless you resubscribe to it.

Reviewer: ```suggestion
You can subscribe to a topic that was not previously published in Salesforce. However, when the topic is published, you do not receive notifications regarding that topic unless you resubscribe to it.
```

[[topicrep]]
=== Replay a Topic

A subscriber can specify which events to receive. By default, a subscriber receives only the events that occur after subscribing. Events outside the 24-hour retention period are discarded.

The Replay Topic Listener operation provides these options:

* `ALL`
+
Subscriber receives all events, including past events that are within the 24-hour retention period and new events sent after the client subscribes.
* `ONLY_NEW`
+
Subscriber receives new events that are broadcast after the client subscribes.
* `FROM_REPLAY_ID`
+
Subscriber receives all events after the specified event `replayId`.

> If you specify the `ALL` or `ONLY_NEW` replay option, the `replayId` value is ignored.

Reviewer: ```suggestion
If you specify either the `ALL` replay option or the `ONLY_NEW` replay option, the `replayId` value is ignored.
```

> In Studio, the *Resume from the Last Replay Id* check box enables you to specify an automatic replay of stored events, based on the Replay ID of the last event processed by the connector. You can use this functionality when the connector stops listening, such as a during a server shutdown or dropped connection. If the stored Replay ID is outside the 24-hour retention period, the replay option determines what events to replay.

Reviewer: check box > checkbox
GLOBAL, per Term Presentation List

[[pushdata]]
== Push Data to Salesforce

You can use Salesforce Connector as an outbound connector in your flow to push data to Salesforce. To use the connector in this capacity, place Salesforce Connector to the right of an inbound endpoint in your flow.

[[receivedata]]
== Receive Data From Salesforce

You can use Salesforce Connector as an inbound connector to stream data from Salesforce into your application. To use the connector in this capacity, use one of these input sources:

* Subscribe topic listener
* Subscribe channel listener
* Replay topic listener
* Replay channel listener

Studio automatically converts the connector to Salesforce Streaming mode, which listens to notifications from the Salesforce Streaming API.

In this example, Salesforce Connector listens to notifications on a topic and feeds the data into the flow:

image::salesforce/salesforce-studio-subscribe-streaming-channel.png[subscribe streaming channel]

> Streaming channels provide notifications to subscribers that are not limited to record-based events. You can use Salesforce Connector to work with Salesforce streaming channels.

Reviewer: ```suggestion
Streaming channels provide subscribers with notifications that are not limited to record-based events. You can use Salesforce Connector to work with Salesforce streaming channels.
```

[[streamsub]]
== Subscribe to a Streaming Channel

> After you create a streaming channel, you can start receiving events by subscribing to the channel. The `subscribe-channel-listener` acts like an inbound endpoint. In this example, every time a subscription to `/u/TestStreaming` receives an event, it executes the rest of the flow and logs a message at the INFO level:

Reviewer: What is the common noun object for "the `subscribe-channel-listener`"?

</flow>
----

The Streaming channel field of the Subscribe channel listener operation does not display change events that are available in the Salesforce environment, because this information is not provided in the Salesforce API. However, your connector can subscribe to a streaming channel to obtain this information. For example, to subscribe to the `All Change Events` channel, use `/data/ChangeEvents` as the channel name to which you subscribe.

For more information, see https://developer.salesforce.com/docs/atlas.en-us.change_data_capture.meta/change_data_capture/cdc_subscribe_channels.htm[Subscription Channels] in the Salesforce Change Data Capture Developer Guide.

=== Streaming Channel Inbound Properties

This information gets passed along as inbound properties:

* `channel` maps to the Channel JSON property.
* `type` maps to the Type JSON property in data.
* `createdDate` maps to the createdDate JSON property in data.

Except for `channel`, each property inside an event is available as an inbound property.

> // Kim, I have a question in to the developer about the above text, since I find it confusing.

Reviewer: Good call!

=== Replay Events from a Streaming Channel

> A streaming channel can replay notifications. The `replay-channel-listener` input source acts like an inbound endpoint and you can use it as shown in the following example:

Reviewer: ```suggestion
A streaming channel can replay notifications. The `replay-channel-listener` input source acts as an inbound endpoint. You can use it as shown in the following example:
```

</flow>
----

> If you specify the `ALL` or `ONLY_NEW` replay options, the `replayId` value is ignored.

Reviewer: ```suggestion
If you specify either the `ALL` replay option or the `ONLY_NEW` replay option, the `replayId` value is ignored.
```

=== Push Events to a Streaming Channel

> Salesforce lets you push custom events to a specific streaming channel through the REST API. To do this, use https://workbench.developerforce.com/about.php[workbench] or this connector.

Reviewer: ```suggestion
Salesforce enables you to push custom events to a specific streaming channel through the REST API. To do this, use https://workbench.developerforce.com/about.php[Workbench] or this connector.
```

[[batchdata]]
== Load Data in Batches

The Salesforce Bulk API loads batches of your organization's data into Salesforce. Salesforce Connector provides the Create and Create Bulk operations for working with the Bulk API.

For all bulk operations, Salesforce handles the creation process in the background. The connector replies with a BatchInfo object, which contains the ID of the batch and the ID of the job it creates to upload the batched objects.

> In Salesforce, you can track the status of bulk data load jobs and their associated batches by using the job ID for the Bulk Data Load Jobs:

Reviewer: ```suggestion
In Salesforce, you can track the status of bulk data load jobs and their associated batches by using the job ID for Bulk Data Load Jobs:
```

. Click YOUR_NAME > *Setup* > *Monitoring* > *Bulk Data Load Jobs*.
. Click the job ID to view the job detail page.

> The job detail page includes a related list of all the batches for the job. The related list provides `View Request` and `View Response` links for each batch. If the batch is a CSV file, the links return the request or response in CSV format. If the batch is an XML file, the links return the request or response in XML format.

Reviewer: What is "a related list"? Do we mean "The job detail page includes a list of all the batches related to the job."?

[[usagenotes]]
== Usage Notes

=== Upsert

> Unless you configure the External ID Field Name for the sObject to which you are trying to upsert, the Upsert operation fails.

Reviewer: ```suggestion
Unless you configure External ID Field Name for the sObject to which you are trying to upsert, the Upsert operation fails.
```

> The Upsert operation does not work with the sObject `priceBookentry2`.

Reviewer: What is the common noun object for "the sObject `priceBookentry2`"?

Author: This is a Salesforce API term. I am leaving as is.

> Although you can't change the `contentType` for a bulk upsert, you can use the Create job operation to set the contentType to either CSV or zipped CSV (if you're near the character limit). Follow up with the Create batch operation.

Reviewer: ```suggestion
Although you can't change the `contentType` value for a bulk upsert, you can use the Create job operation to set the content type to CSV (or zipped CSV if you're near the character limit). Follow up with the Create batch operation.
```

=== Query

Although you can see the fields of an sObject and their corresponding types via DataSense, the Query operation returns all fields as `String`.

To use the actual type of a field, use a Transform or Transform Message component to convert the field to the desired type.

> Although `CreatedDate` field appears as `dateTime`, the query returns a `String` that represents the date. To use the field as a `dateTime` field, configure it using a Transform Message component.

Reviewer: ```suggestion
Although the `CreatedDate` field appears as `dateTime`, the query returns a `String` value that represents the date. To use the field as a `dateTime` field, configure it using a Transform Message component.
```

		</ee:transform>
----

In this example:

* The `body` element contains `URLParameters`, which is a map containing the parameters that replace the wildcards in the path of the REST resource described in the Apex class.
+

> For example if the REST resource is set to `@RestResource(urlMapping='/myResource/\*/mySubResource/*')`, the value of `Parameter1` replaces the first `\*`, and the value of `Parameter2` replaces the second `*`.

Reviewer: ```suggestion
For example if the REST resource is set to `@RestResource(urlMapping='/myResource/\*/mySubResource/*')`, the value of `Parameter1` replaces the first `\*` and the value of `Parameter2` replaces the second `*`.
```

* Key names must start with `Parameter`, followed by a number that shows the position of the `*` to be replaced.
+

> * After the `URLParameters` block, provide the content of the `body` to send to the REST resource, as shown in the example `account` block.

Reviewer: ```suggestion
* After the `URLParameters` block, provide the content of the `body` value to send to the REST resource, as shown in the example `account` block.
```

* The `headers` and `cookies` fields describe the headers and cookies to pass along with the HTTP request to the desired service.
* The `queryParameters` field describes the query parameters to use, and the keys and values in this map that the specified Apex Class must accept.

=== Insert Values in a Salesforce Drop-Down Menu

Inserting dependent values into an existing drop-down menu in Salesforce does not always work. Test to confirm this functionality.

=== Evaluate Values in a Salesforce Drop-Down Menu

> If you are evaluating against a value in an existing drop-down list field in Salesforce, use the exact value in the drop-down list. For example, if you use the value `US` to evaluate against the contents of a drop-down list that contains the value `USA`, the evaluation works, but the result is two values in the drop-down list: one for `US` and one for `USA`.

Reviewer: GLOBALLY evaluate whether the audience requires us to specify "drop-down" for lists and menus. Even I understand what the little downward-facing arrow means!  lol

Reviewer: In addition to the above recommendation (not reflected here)...:
```suggestion
If you are evaluating against a value in an existing drop-down list field in Salesforce, use the exact value in the list. For example, if you use the value `US` to evaluate against the contents of a drop-down list that contains the value `USA`, the evaluation works, but the result is two values: one for `US` and one for `USA`.
```

=== Limits on API Calls

> You must know the rate limiting policy that applies to your account so that you do not exceed the number of allotted API calls per day.

Reviewer: ```suggestion
You must know the parameters of the rate-limiting policy that applies to your account so that you do not exceed the number of allotted API calls per day.
```

Author: Changing to "scope of."

=== Opportunity Object

> When extracting data from an Opportunity, be aware that a "quarter" in this context is relative to the financial year of the organization instead of relative to the calendar year.

Reviewer: ```suggestion
When extracting data from an Opportunity object, be aware that a "quarter" in this context is relative to the financial year of the organization and not necessarily to the calendar year.
```

Author: This is a Salesforce object. I qualified it as such
